### PR TITLE
Proposed interface changes to support async + explainability + analysis

### PR DESCRIPTION
**Pull request by ellis-bigelow**
# Interface Proposal
## Background
Given the recent progress towards explainability and impending exploration of payload logging, skew, bias, and outlier detection, I think it's time to make some refinements to our interface. The earlier we align on these interfaces, the sooner we can make any necessary backwards incompatible changes, minimizing customer impact.

_Thanks to Clive (this is largely influenced by his work with MLGraph) and Bhupesh for brainstorming some of these ideas with me._

### Goals ### 
1. Maximize the simplicity of the spec, both in the MVP and Bells and Whistles cases.
2. Provide an excellent out-of-the-box experience for all features, with configuration overrides if necessary.
3. Cover 80% of use cases, KFServing cannot solve all ML workloads.

### High Level Ideas ###
1. "Canarying" only matters for Models, not explainers.
2. Skew, Bias, Outliers, and Logging can be consolidated into an "Analysis" concept.
3. "Custom" should be duck typed to Pod Template.

## Canarying ## 
As we expand the spec, we have to decide if all components of the spec are canaryable or not. After some consideration, I'm starting to think that canarying should only apply to the predictor, preprocessor, and postprocessor. That is to say, logging configuration, skew detection, and explainers are siblings of `default` and `canary` and cannot themselves be canaried.

At a high level we're choosing between:
``` 
spec.default.tensorflow: ...
spec.default.preprocessor: ...
spec.canary.tensorflow: ...
spec.canary.preprocessor: ...
spec.explain.alibi: ...
spec.analysis.skew: ...
```
and 
```
spec.default.prediction.tensorflow: ...
spec.default.preprocessor: ...
spec.default.explanation.alibi: ...
spec.default.analysis.skew: ...
spec.canary.prediction.tensorflow: ...
spec.canary.preprocessor: ...
spec.canary.explanation.alibi: ...
spec.canary.analysis.skew: ...
```
The second approach is clearly more verbose, but more importantly, it isn't clear what the use case for canarying an explainability config or payload logging config. 

The primary usecase of model canarying is to run a new revision of a model under a percentage of traffic for validation. For skew and bias, these algorithms are statistical in nature, and it is not clear what the evaluation path for changes to these configurations would be. That is to say, there should be no way to break a KFService with an `AnalysisSpec` change, so what's the point of including it as a child of the `CanarySpec`?

**After all of these learnings, it should be on the table to discuss whether or not canarying should be moved to a higher abstraction (i.e. user controlled or managed with an MLGraph or Route CRD). That said, I do still feel it is useful in KFService.**

## Analysis ## 
Previously, we've been thinking of all the concepts as separate units of a static graph (see KFServing data plane architecture diagram). Instead, it may make sense to consolidate some of these concepts into an Analysis concept. The primary benefit of separating these configs from the predictor is to enable different scaling of these components. If each feature we include results in a new top level key of the KFService as well as a new Knative Configuration for the feature, we risk our users viewing both our interface and implementation as bloated.

Skew, Bias, and Outliers are all statistical concepts that should have first class support in KFServing. A user should not need to supply their own image -- instead, configuration of these feature should be driven through the CRD. There may be some cases where cloud providers (e.g. Seldon Deploy or AzureML or Cloud AI Platform) will want to override these implementations, but this should be done from a Configmap, not through the Knative interface (i.e., they are operator-level details).

It makes sense that `logging` is also included in `analysis`, as the purpose of payload logging is for offline analysis. Along those lines, it makes sense that different features of analysis may or may not run  online or offline. At the end of the day, this component should be a request+response processor, and any feature that makes sense for this context should execute here and be configurable via the KFService CRD API.

## Pod Template Ducktyping ##
This one is far less controversial. Knative is moving to a limited PodTemplateSpec that is ducktyped to the Kubernetes DeploymentTemplateSpec. This spec should be used to support Custom Explainers and Custom Predictors in KFServing. 

That said, we should continue to drive users towards our out-of-the-box solutions for prediction and explainability, as it allows them to focus on their ML workloads, and not on kubernetes details.

As with Knative, we will support a limited set of the PodTemplateSpec where necessary.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/179)

### CODE REVIEW

**Comment by bhupc** [comment]
I don't have a ton of experience on explainability, so some one with more exp there should weigh in. For some well known frameworks like TF, sklearn, the explainers might be well known and may be not part of the model building workflow. 
But does the same argument hold for a custom container model? would this be a custom explainer? And would that canarying be really painful then with KFServing?  We are essentially saying that custom explainer should be canaried and tested elsewhere similar to the logging and other analysis parts. 
That can be fair. Explainer doesn't have to be "in-sync" with the model.

**Comment by cliveseldon** [comment]
I think this looks good.

I see you separated explanation from skew, bias and outliers (in analysis). Is this because explanation is of interest to the model user directly but skew, bias, outliers are more of interest to data scientists and dev-ops so would never be returned inline but always to some other endpoint for offline analysis and alerting to model creators?

**Comment by ellis-bigelow** [comment]
> inability, so some one with more exp there should weigh in. For some well known frameworks like TF, sklearn, the explainers might be well known and may be not part of the model building workflow.
> But does the same argument hold for a custom container model?

Very succinct :). This makes sense to me, but I agree would should get more +1s.

**Comment by ellis-bigelow** [comment]
> model user directly but skew, bias, outliers are more of interest to data scientists and dev-ops so would never be returned inline

**Comment by ellis-bigelow** [comment]
> is this because explanation is of interest to the model user directly but skew, bias, outliers are more of interest to data scientists and dev-ops so would never be returned inline but always to some other endpoint for offline analysis and alerting to model creators?

I separated it primarily due to "At the end of the day, this component should be a request+response processor, and any feature that makes sense for this context should execute here and be configurable via the KFService CRD API."

The explainer is much more involved and can't be boiled down as much as the other analysis tools. My bet is that the analysis tools will only be out of the box, where users may need to provide their own explainer image. 

Regarding online or offline, the Analyzer may do online skew detection and async payload logging to bigquery in the same component.

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 1]
```diff
@@ -79,9 +91,35 @@ type SKLearnSpec struct {
...
+
+// AnalysisSpec defines the configuration of the analyzer default configuration to route traffic.
+type AnalysisSpec struct {
+	Skew    *SkewSpec    // TODO
+	Bias    *BiasSpec    // TODO
+	Logging *LoggingSpec // TODO
+	// TODO Include configuration of backend datastore. KFServing should provide an image
+}
+
+type SkewSpec struct {
```
Comment (cliveseldon) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: Should we add an OutlierSpec as well? And if so this does seem closer to ExplanationSpec where there are multiple algorithms that fit different types of models better so the user might want  control based on their model.

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 2]
```diff
@@ -79,9 +91,35 @@ type SKLearnSpec struct {
 	Resources v1.ResourceRequirements `json:"resources,omitempty"`
 }
 
-// CustomSpec provides a hook for arbitrary container configuration.
-type CustomSpec struct {
-	Container v1.Container `json:"container"`
+// ExplanationSpec defines the configuration of the model explainer.
+type ExplanationSpec struct {
+	Async bool `json:"omitempty"` // Defaults false. If true, creates and outputs to a well known (named) Knative Channel
```
[the remark this answers is not in the document]
Comment (ellis-bigelow) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: IMO, we should just emit to something well known and derived from the kfservice's name, similar to how the uri works. @yuzisun, do you agree with this given your experience with knative eventing?

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (ellis-bigelow) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: I'm a little confused. Can you explain the outlier usecase? My understanding is that it's just a statistical definition?

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuzisun) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: Is this flag for the output of the explainer ? Ye I agree we can create a channel which is derived from kfservice name, then explainer is responsible for delivering the result to this channel.

[the remark this answers is not in the document]
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: How will the user have configure the channel provisionor to be inmemory or kafka or NATS?

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: There are various [online outlier detection algorithms] we have been working on at Seldon(https://github.com/SeldonIO/seldon-core/tree/master/components/outlier-detection) which need to be matched to the model. So I see it as a similar case to explainability where you want to configure which you use.

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuzisun) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: This can be in the config map ?

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (yuzisun) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: Does skew here mean for the data distribution change or shift ?

Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: Skew (or concept drift) would be identifying a decrease in accuracy over a time period/batch of requests while outlier detection would be for flagging individual requests as outliers to the current distribution. @jklaise @arnaudvl ?

Comment (jklaise) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: @cliveseldon correct, outlier detection needs to be tied to the specific example as an outlier detection is a model telling wether a particular data point is an anomaly (could be wrt to training data or wrt to samples seen so far in an online updating fashion), so it's similar to the explainer use case, while concept drift monitors the data stream as a whole and produces statistics/alarms (e.g. detecting a change point in the data stream).

Comment (ellis-bigelow) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: Interesting. It's almost starting to sound like outlier detection shouldn't happen within a KFService. Now I'm questioning whether or not explainability should happen within a KFService. Can we break this down into a "Whitebox" vs "Blackbox" problem, where anything that's blackbox can have OOTB solutions that enhance your model, and anything whitebox is considered custom and must be deployed outside your model.

**Comment by rakelkar** [comment]
> The second approach is clearly more verbose, but more importantly, it isn't clear what the use case for canarying an explainability config or payload logging config.

You need to be able to safely rollout any component that matters! Safe rollout doesn't feel like an optional feature to me?

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: I think the key is more components which are online (they listen to requests/responses and build internal models) and are (optionally) part of the request/response flow - add an explanation, outlier detection alert, concept drift alert etc. They need to be associated and configured with a particular unit of prediction.

Comment (ellis-bigelow) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: This is where we landed thursday night. Many of the analysis components get pushed out (everything offline), and there's a small set of online components for a static graph -- somewhat similar to our original stance.

**Comment by ellis-bigelow** [comment]
> > The second approach is clearly more verbose, but more importantly, it isn't clear what the use case for canarying an explainability config or payload logging config.
> 
> You need to be able to safely rollout any component that matters! Safe rollout doesn't feel like an optional feature to me?

Yeah I'm pretty convinced of this again.

**Comment by ellis-bigelow** [comment]
/hold

this is quite out of date due to discussion.
